### PR TITLE
Surveys [WIP]

### PR DESCRIPTION
**Pull request by mattandrews**
This will still change (likely visually) but probably worth getting your eyeballs on it in its current state @davidrapson.

Quite a lot here really and a few `@TODO`s to clean up by me, but initial thoughts welcome.

### CODE REVIEW

File: modules/boilerplate/getSurveys.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+'use strict';
+const _ = require('lodash');
+const models = require('../../models/index');
+
+// @TODO fetching new surveys requires an app restart... do we care?
```
Comment (mattandrews) [inline modules/boilerplate/getSurveys.js]: Yes we do care – will change this!

Comment (mattandrews) [reply modules/boilerplate/getSurveys.js]: ... would it be better to check for new/updated surveys every x minutes, or to simply make the query on every pageload? Or stick to as-is right now (eg. they're retrieved at app start time)?

**Comment by davidrapson** [comment]
Probably worth chatting through this in person, but a thought:

Should we consider this entirely a progressive enhancement? If so, we could fetch the surveys client-side via a cached endpoint rather than needing to couple the rendering of the app to fetching surveys. Would also bypass the problem of restarting the app to get new surveys or getting into the territory of worker queues to fetch them.

File: assets/js/modules/surveys.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
+'use strict';
+
+const $ = require('jquery');
+
+// via https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API
+let storageAvailable = type => {
```
Comment (davidrapson) [inline assets/js/modules/surveys.js]: Maybe pop this in a util file?

**Comment by mattandrews** [comment]
Yeah, good point – so just make an AJAX call on pageload to check if the current page has a survey, that kind of thing?

File: assets/js/modules/surveys.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
+/* global Vue */
+'use strict';
+
+const $ = require('jquery');
+
+// via https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API
+let storageAvailable = type => {
```
Comment (davidrapson) [inline assets/js/modules/surveys.js]: Should this go in a utils file?

File: assets/sass/components/survey.scss [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            margin-bottom: 0;
+        }
+
+        .survey__submit {
+            flex-grow: 0;
+            margin-bottom: 10px; // fixes weird alignment issue
+        }
+    }
+
+    label {
```
Comment (davidrapson) [inline assets/sass/components/survey.scss]: Any reason these elements don't have BEM-style class names for consistency?

File: assets/sass/utilities/utilities.scss [hunk 5]
```diff
@@ -31,6 +31,13 @@
     display: none;
 }
 
+// hide elements for users with JS
+.js-on {
+    .u-hidden-for-js {
```
Comment (davidrapson) [inline assets/sass/utilities/utilities.scss]: Does this need to be nested?

File: controllers/toplevel/index.js [hunk 6]
```diff
@@ -296,8 +298,9 @@ module.exports = (pages, sectionPath, sectionId) => {
     // handle contrast shifter
     router.get('/contrast/:mode', (req, res) => {
         res.cacheControl = { maxAge: 1 };
-        let duration = 6 * 30 * 24 * 60 * 60; // 6 months
+
         let cookieName = config.get('cookies.contrast');
+        let duration = moment.duration(6, 'months').asMilliseconds();
```
Comment (davidrapson) [inline controllers/toplevel/index.js]: 👍 Nice

File: controllers/toplevel/index.js [hunk 7]
```diff
@@ -310,6 +313,113 @@ module.exports = (pages, sectionPath, sectionId) => {
         res.redirect(redirectUrl);
     });
 
+    // https://stackoverflow.com/questions/20053949/how-to-get-the-google-analytics-client-id
```
Comment (davidrapson) [inline controllers/toplevel/index.js]: Is this comment connected? 🤔

File: controllers/toplevel/index.js [hunk 8]
```diff
@@ -310,6 +313,113 @@ module.exports = (pages, sectionPath, sectionId) => {
...
+        body('choice')
+            .exists()
+            .not()
+            .isEmpty()
+            .isInt()
+            .withMessage('Please supply a valid choice')
+    ];
+
+    // @TODO add this endpoint to cloudfront
+    router.post('/get-survey', (req, res) => {
```
Comment (davidrapson) [inline controllers/toplevel/index.js]: As discussed, lets make this a cached GET endpoint.

File: modules/boilerplate/middleware.js [hunk 9]
```diff
@@ -21,6 +21,8 @@ const routes = require('../../controllers/routes');
 // load auth strategy
 require('../../modules/boilerplate/auth');
 
+let sessionSecret = getSecret('session.secret') || process.env.sessionSecret;
```
Comment (davidrapson) [inline modules/boilerplate/middleware.js]: Should this be the other way around, with env first?

File: test/features/survey.test.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+let testSurveyData = {
+    name: 'Test Survey',
+    question_en: 'What is the meaning of life?',
+    question_cy: 'Beth yw ystyr bywyd?',
+    activePath: '/foo',
+    active: 1,
+    choices: [
+        {
+            title_en: '42',
```
Comment (davidrapson) [inline test/features/survey.test.js]: :neckbeard:

File: bin/cloudfront/test.json [hunk 11]
```diff
@@ -2887,6 +2887,129 @@
                 "Items": [],
                 "Quantity": 0
             },
+            "QueryString": false
+        },
+        "MaxTTL": 31536000,
```
Comment (davidrapson) [inline bin/cloudfront/test.json]: This diffs are genuinely nice 😍

**Comment by davidrapson** [comment]
:shipit:

**Comment by mattandrews** [comment]
I thought about filtering on the client but preferred keeping the URL rewriting (eg. treating `/welsh/contact` and `/contact` as the same thing) on the server-side where it happens elsewhere too.

**Comment by davidrapson** [comment]
"…but preferred keeping the URL rewriting (eg. treating /welsh/contact and /contact as the same thing) on the server-side where it happens elsewhere too."

Good point. 👍

**Comment by davidrapson** [comment]
👍  New commits look 🆗

**Pull request merged**